common/utils/include/claragenomics/utils/graph.hpp

```diff
@@ -0,0 +1,300 @@
+/*
+* Copyright (c) 2019, NVIDIA CORPORATION.  All rights reserved.
+*
+* NVIDIA CORPORATION and its licensors retain all intellectual property
+* and proprietary rights in and to this software, related documentation
+* and any modifications thereto.  Any use, reproduction, disclosure or
+* distribution of this software and related documentation without an express
+* license agreement from NVIDIA CORPORATION is strictly prohibited.
+*/
+
+#pragma once
+
+#include <vector>
+#include <unordered_map>
+#include <unordered_set>
+#include <stdint.h>
+#include <string>
+#include <iostream>
+#include <sstream>
+#include <algorithm>
+
+namespace claragenomics
+{
+
+using node_id_t     = int32_t;
+using edge_weight_t = int32_t;
+using edge_t        = std::pair<node_id_t, node_id_t>;
+
+/// \struct PairHash
+/// Hash function for a pair
+struct PairHash
+{
+public:
+    /// \brief Operator overload to define hash function
+    template <class T1, class T2>
+    size_t operator()(const std::pair<T1, T2>& pair) const
+    {
+        size_t hash_1 = std::hash<T1>()(pair.first);
+        size_t hash_2 = std::hash<T2>()(pair.second);
+        return hash_1 ^ hash_2;
+    }
+};
+
+/// \brief Object representing a generic graph structure
+class Graph
```

Comment:
Only use the Directed and Undirected classes directly. Sure, will get rid of the ctor and dtor for Graph. But since the the derived classes need the serialize and add_edge functions to be custom, wouldn't they need `virtual`? Or did you mean all except those?

Reply:
Since they don't provide any functionality in the base class and you don't intend a runtime interface just remove them and only define them in the derived classes.
Even if you had implemented them, there is no technical reason to make them virtual as long as you don't want to refer to the derived classes by a common `Graph* g = new DirectedGraph();` (or similar).
The only "problem" if you have the same non-virtual function in a base and derived class is function/name hiding, i.e. the DirectedGraph function will hide ALL the base class' function with the same name (unless you do `using Base::function;` explicitly) - which can lead to nasty bugs if you are not careful.